concepts/websockets.textile

```diff
@@ -0,0 +1,591 @@
+
+There has been a little "controversy":https://www.reddit.com/r/node/comments/91kgte/uws_has_been_deprecated/ around μWS recently due to the author having attempted to pull it from NPM for philosophical reasons, but "the latest working version":https://www.npmjs.com/package/uws/v/10.148.1 remains on NPM and can be installed specifying that version explicitly when installing from NPM. That said, the author is working on "a new version":https://github.com/uNetworking/v0.15, with accompanying "node.js bindings":https://github.com/uNetworking/uWebSockets-node also "in development":https://github.com/uNetworking/uWebSockets-node/issues/2.
+
+h4(#minimal-code-sample). Minimal code sample
+
+*Server (Node.js):*
+
+```[js]
+var WebSocketServer = require('uws').Server;
+var wss = new WebSocketServer({ port: 3000 });
+ 
+function onMessage(message) {
+  console.log('received: ' + message);
+}
+ 
+wss.on('connection', function(ws) {
+  ws.on('message', onMessage);
+  ws.send('something');
+});
+```
+
+h3(#faye-websocket). faye-websocket
+
+"faye-websocket":https://github.com/faye/faye-websocket-node is a standards-compliant WebSocket implementation for both the client and server, and originated from the Ruby-on-Rails community as part of the "Faye project":https://faye.jcoglan.com/. As per the Github, project:
+
+bq. It does not provide a server itself, but rather makes it easy to handle WebSocket connections within an existing Node application. It does not provide any abstraction other than the standard WebSocket API.
+
+
+
+In the sample code below for the server, you can see that all of the work of handling the connection upgrade and translating messaging frames from inbound socket buffers is handled by the @WebSocket@ class provided by the library. As with other minimal solutions, this is a no-frills implementation - you'll need to handle application-specific concerns yourself.
+
+h4(#minimal-code-samples-1). Minimal code samples
+
+*Client (Browser, before bundling):*
+
+```[js]
+var WebSocket = require('faye-websocket'),
+    ws        = new WebSocket.Client('ws://www.example.com/');
+
+ws.on('open', function(event) {
+  console.log('open');
+  ws.send('Hello, world!');
+});
+
+ws.on('message', function(event) {
+  console.log('message', event.data);
+});
+
+ws.on('close', function(event) {
+  console.log('close', event.code, event.reason);
+  ws = null;
+});
+```
+
+*Server (Node.js):*
+
+```[js]
+var WebSocket = require('faye-websocket'),
+    http      = require('http');
+
+var server = http.createServer();
+
+server.on('upgrade', function(request, socket, body) {
+  if (WebSocket.isWebSocket(request)) {
+    var ws = new WebSocket(request, socket, body);
+    
+    ws.on('message', function(event) {
+      ws.send(event.data);
+    });
+    
+    ws.on('close', function(event) {
+      console.log('close', event.code, event.reason);
+      ws = null;
+    });
+  }
+});
+
+server.listen(8000);
+```
+
+h3(#socket.io). Socket.io
```

Comment:
dot in the link messes up the heading. just use socketio

Reply:
Fixed.